go/chat/unfurl/scrape_generic.go

```diff
@@ -9,58 +9,198 @@ import (
+	defaultFaviconURL, err := GetDefaultFaviconURL(uri)
+	if err != nil {
+		return res, err
+	}
+	generic.setFaviconURL(&defaultFaviconURL, 0)
+
 	hostname, err := GetHostname(uri)
 	if err != nil {
 		return res, err
 	}
-	generic.Url = uri
-	generic.SiteName = domain
+
 	c := colly.NewCollector()
+	// scrape open graph tags
 	c.OnHTML("head meta[content][property]", func(e *colly.HTMLElement) {
-		prop := e.Attr("property")
-		content := e.Attr("content")
+		score := getOpenGraphScore(domain)
+		prop := strings.ToLower(e.Attr("property"))
+		content := strings.Trim(e.Attr("content"), " ")
 		switch prop {
-		case "og:description":
-			generic.Description = &content
-		case "og:image":
-			generic.ImageUrl = new(string)
-			*generic.ImageUrl = fullURL(hostname, content)
+		case "og:title":
+			generic.setTitle(content, score)
+		case "og:url":
+			generic.setURL(content, score)
 		case "og:site_name":
-			generic.SiteName = content
+			generic.setSiteName(content, score)
+		case "og:image":
+			imageURL := fullURL(hostname, e.Attr("href"))
+			generic.setImageURL(&imageURL, score)
+			imageURL = fullURL(hostname, content)
+			generic.setImageURL(&imageURL, score)
 		case "og:pubdate":
-			s.Debug(ctx, "pubdate: %s", content)
-			t, err := time.Parse("2006-01-02T15:04:05Z", content)
-			if err == nil {
-				generic.PublishTime = new(int)
-				*generic.PublishTime = int(t.Unix())
-			} else {
-				s.Debug(ctx, "scrapeGeneric: failed to parse pubdate: %s", err)
-			}
+			s.setAndParsePubTime(ctx, content, generic, score)
+		case "og:description":
+			generic.setDescription(&content, score)
 		}
 	})
+
+	// scrape twitter/non open graph
+	c.OnHTML("head meta[content][name]", func(e *colly.HTMLElement) {
+		score := getTwitterScore(domain)
+		name := strings.ToLower(e.Attr("name"))
+		content := strings.Trim(e.Attr("content"), " ")
+		switch name {
+		case "twitter:title":
+			generic.setTitle(content, score)
+		case "twitter:image":
+			imageURL := fullURL(hostname, e.Attr("href"))
+			generic.setImageURL(&imageURL, score)
+			imageURL = fullURL(hostname, content)
+			generic.setImageURL(&imageURL, score)
+		case "twitter:description":
+			generic.setDescription(&content, score)
+		case "application-name":
+			generic.setSiteName(content, defaultScore)
+		case "description":
+			generic.setDescription(&content, defaultScore)
+		case "pubdate":
+			s.setAndParsePubTime(ctx, content, generic, defaultScore)
+		case "lastmod":
+			s.setAndParsePubTime(ctx, content, generic, defaultScore)
+		}
+	})
+
+	// scrape title
 	c.OnHTML("head title", func(e *colly.HTMLElement) {
-		generic.Title = e.Text
+		generic.setTitle(e.Text, defaultScore)
 	})
+
+	// scrape favicon
 	c.OnHTML("head link[rel][href]", func(e *colly.HTMLElement) {
 		rel := strings.ToLower(e.Attr("rel"))
-		if strings.Contains(rel, "shortcut icon") {
-			generic.FaviconUrl = new(string)
-			*generic.FaviconUrl = fullURL(hostname, e.Attr("href"))
+		if strings.Contains(rel, "shortcut icon") ||
```

Comment:
Could you add scores for these? Some of these are higher res (like `apple-touch-icon` (I would prefer the higher res ones).

Reply:
sgtm

Reply:
Note this is an example why I think per attribute scores make sense

Reply:
Note this is an example why I think per attribute scores make sense